tinygrad/codegen/linearizer.py

```diff
@@ -471,11 +471,16 @@ def uop(self, uop:UOps, dtype:Optional[DType], vin:Tuple[UOp, ...], arg:Any=None
     if cachable: self.saved_exprs[key] = ret
     return ret
 
+  def parse_cast(self, uop: UOp, dtype:DType, bitcast:bool) -> UOp:
+    if isinstance(dtype, ImageDType): return uop
+    if uop.uop == UOps.LOAD and uop.vin[0].dtype != uop.dtype: uop = self.uop(UOps.CAST, uop.vin[0].dtype, (uop,), (uop.vin[0].dtype, False))
```

Comment:
sorta a hack to go around buf dtype being loaded as fp32, appends a CAST to the end of the AST in the middle of a LOAD<>CAST that casts the fp32ed load back to its original dtype, this will fix that nan probablities in llama

Reply:
not relevant anymore, doing this the right way